backend/cfg/cfg_dataflow.ml

```diff
@@ -132,3 +132,192 @@ module Forward (D : Domain) (T : Transfer with type domain = D.t) :
+    (* note: `handler_map` contains the value at the *start* of the block. *)
+    let handler_map : D.t Label.Tbl.t =
+      Label.Tbl.create (Label.Tbl.length cfg.Cfg.blocks)
+    in
+    while (not (WorkSet.is_empty !work_set)) && !iteration < max_iteration do
+      incr iteration;
+      let element, block = remove_and_return cfg work_set in
+      let exn : domain =
+        Option.map
+          (fun exceptional_successor ->
+            Label.Tbl.find_opt handler_map exceptional_successor)
+          block.exn
+        |> Option.join
+        |> Option.value ~default:D.bot
+      in
+      let value = transfer_block element.value ~exn block in
+      if block.is_trap_handler
+      then begin
+        let old_value =
+          Option.value
+            (Label.Tbl.find_opt handler_map block.start)
+            ~default:D.bot
+        in
+        let new_value = T.exception_ value in
+        if not (D.less_equal new_value old_value)
+        then begin
+          Label.Tbl.replace handler_map block.start new_value;
+          List.iter
+            (fun predecessor_label ->
+              let current_value =
+                Option.value
+                  (Label.Tbl.find_opt res predecessor_label)
+                  ~default:D.bot
+              in
+              work_set
+                := WorkSet.add
+                     { WorkSetElement.label = predecessor_label;
+                       value = current_value
+                     }
+                     !work_set)
+            (Cfg.predecessor_labels block)
+        end
+      end
+      else
+        List.iter
+          (fun predecessor_label ->
+            let old_value =
+              Option.value
+                (Label.Tbl.find_opt res predecessor_label)
+                ~default:D.bot
+            in
+            let new_value = D.join old_value value in
+            if not (D.less_equal new_value old_value)
+            then begin
+              Label.Tbl.replace res predecessor_label new_value;
+              let already_in_workset = ref false in
+              work_set
+                := WorkSet.filter
+                     (fun { WorkSetElement.label; value } ->
+                       if Label.equal label predecessor_label
+                       then begin
+                         if D.less_equal new_value value
+                         then already_in_workset := true;
+                         not (D.less_equal value new_value)
+                       end
+                       else true)
+                     !work_set;
+              if not !already_in_workset
+              then
+                work_set
+                  := WorkSet.add
+                       { WorkSetElement.label = predecessor_label;
+                         value = new_value
+                       }
+                       !work_set
+            end)
+          (Cfg.predecessor_labels block)
+    done;
+    if !iteration < max_iteration then Result.Ok res else Result.Error res
```

Comment:
This condition should also check if the set is empty, same as the while loop, and the same for forward analysis. Otherwise, it returns `Result.Error` instead of Ok when the fixpoint reached in the last iteration before max_iteration.